urllib3/_collections.py

```diff
@@ -227,20 +227,20 @@ def add(self, key, val):
                 # Need to convert the tuple to list for further extension
                 _dict_setitem(self, key_lower, [vals[0], vals[1], val])
 
-    def extend(*args, **kwargs):
+    def extend(self, *args, **kwargs):
         """Generic import function for any type of header-like object.
         Adapted version of MutableMapping.update in order to insert items
         with self.add instead of self.__setitem__
         """
-        if len(args) > 2:
-            raise TypeError("update() takes at most 2 positional "
+        if len(args) > 1:
+            raise TypeError("extend() takes at most 1 positional "
                             "arguments ({} given)".format(len(args)))
-        elif not args:
-            raise TypeError("update() takes at least 1 argument (0 given)")
-        self = args[0]
-        other = args[1] if len(args) >= 2 else ()
+        other = args[0] if len(args) >= 1 else ()
         
-        if isinstance(other, Mapping):
+        if isinstance(other, HTTPHeaderDict):
```

Comment:
> As long as there is no agreement, how the common header API will look like, it won't make sense to try to detect it imho.

What do you mean? Requests vendors urllib3 inside of its own package. Because it's a *copy*, the two class definitions would not be the same, so isinstance would fail despite that they're effectively the same.

I'm proposing we do more duck-typing rather than instance-checking for internal objects.

> Looks like the only thing we would have to add is something to replicate .lower_items() and then they might be good to go to just take HTTPHeaderDict as it is.

+1.